src/iterator.rs

```diff
@@ -0,0 +1,300 @@
+            return Ok(());
+        }
+
+        self.closed = true;
+        unsafe {
+            archive_result(
+                ffi::archive_read_close(self.archive_reader),
+                self.archive_reader,
+            )?;
+            archive_result(
+                ffi::archive_read_free(self.archive_reader),
+                self.archive_reader,
+            )?;
+        }
+        Ok(())
+    }
+
+    unsafe fn next_header(&mut self) -> ArchiveContents {
+        match ffi::archive_read_next_header(self.archive_reader, &mut self.archive_entry) {
+            ffi::ARCHIVE_EOF => ArchiveContents::EndOfEntry,
+            ffi::ARCHIVE_OK => {
+                let file_name = CStr::from_ptr(ffi::archive_entry_pathname(self.archive_entry))
+                    .to_string_lossy()
+                    .into_owned();
+                ArchiveContents::StartOfEntry(file_name)
+            }
+            _ => ArchiveContents::Err(Error::from(self.archive_reader)),
+        }
+    }
+
+    unsafe fn next_data_chunk(&mut self) -> ArchiveContents {
+        let mut buffer = std::ptr::null();
+        let mut offset = 0;
+        let mut size = 0;
+        let mut target = Vec::with_capacity(READER_BUFFER_SIZE);
+
+        match ffi::archive_read_data_block(self.archive_reader, &mut buffer, &mut size, &mut offset)
+        {
+            ffi::ARCHIVE_EOF => ArchiveContents::EndOfEntry,
+            ffi::ARCHIVE_OK => {
+                let content = slice::from_raw_parts(buffer as *const u8, size);
+                let write = target.write_all(content);
+                if let Err(e) = write {
+                    ArchiveContents::Err(e.into())
+                } else {
+                    ArchiveContents::DataChunk(target)
+                }
+            }
+            _ => ArchiveContents::Err(Error::from(self.archive_reader)),
+        }
+    }
+
+    /// Iterate over the contents of an archive, streaming the contents of each
+    /// entry in small chunks.
+    ///
+    /// ```no_run
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use compress_tools::*;
+    /// use std::fs::File;
+    ///
+    /// let file = File::open("tree.tar")?;
+    ///
+    /// let mut name = String::default();
+    /// let mut size = 0;
+    /// let mut iter = ArchiveIterator::from_read(file)?;
+    ///
+    /// for content in &mut iter {
+    ///     match content {
+    ///         ArchiveContents::StartOfEntry(s) => name = s,
+    ///         ArchiveContents::DataChunk(v) => size += v.len(),
+    ///         ArchiveContents::EndOfEntry => {
+    ///             println!("Entry {} was {} bytes", name, size);
+    ///             size = 0;
+    ///         }
+    ///         ArchiveContents::Err(e) => {
+    ///             Err(e)?;
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// iter.close()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn from_read(source: R) -> Result<ArchiveIterator<R>>
```

Comment:
I think we should move this as the first function on the block. I feel like its always better to have these "constructors" as the first functions on the doc pages.

Reply:
I just did. Please take a look and see if it is good.